File: baselib/src/main/java/com/yanb/daqsoft/baselib/base/BaseActivity.java
package com.yanb.daqsoft.baselib.base;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.yanb.daqsoft.baselib.activities.AppManager;
import com.yanb.daqsoft.baselib.activities.IBasePresenter;
import com.yanb.daqsoft.baselib.activities.IBaseView;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * 功能
 *
 * @author 严博
 * @version 1.0.0
 * @date 2019-6-3.17:14
 * @since JDK 1.8
 */
public abstract class BaseActivity <P extends IBasePresenter> extends AppCompatActivity implements IBaseView {
    protected P presenter;
    public Context mContext;
    public Unbinder mUnbinder;
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mContext = this;
        doBeforeSetContentView();
        setContentView(getLayoutId());
        mUnbinder = ButterKnife.bind(this);
        presenter = initPresenter();
        this.initView();
    }
    // 获取布局ID
    public abstract int getLayoutId();
    // 初始化
    public abstract void initView();
    // 初始化Presenter
    public abstract P initPresenter();

    /**
     * 设置布局前设置
     */
    public void doBeforeSetContentView() {
        // 添加Activity到堆栈
        AppManager.getAppManager().addActivity(this);
    }
    /**
     * 销毁
     */
    @Override
    protected void onDestroy() {
        // 将当前activity移除管理栈
        AppManager.getAppManager().removeActivity(this);
        if (presenter !=null){
            // 在presenter中解绑释放view
            presenter.detach();
            presenter = null;
        }
        if (mUnbinder !=null){
            mUnbinder.unbind();
        }
        super.onDestroy();
    }
}

File: baselib/src/main/java/com/yanb/daqsoft/baselib/ktbase/BaseActivity.kt
package com.yanb.daqsoft.baselib.ktbase

import android.content.Context
import android.os.Bundle
import android.os.PersistableBundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import com.yanb.daqsoft.baselib.view.MultipleStatusView

/**
 * kotlin基础baseActivity
 * @author 严博
 * @date 2019-6-5.16:46
 * @version 1.0.0
 * @since JDK 1.8
 * :继承
 * var: 变量
 * val:常量
 *
 */
abstract class BaseActivity : AppCompatActivity(){
    /**
     * 多布局状态的View切换
     * NULL检查机制
     * Kotlin的空安全设计对于声明可为空的参数，在使用时要进行空判断处理，有两种处理方式，
     * 1、字段后加!!像Java一样抛出空异常，
     * 2、另一种字段后加?可不做处理返回值为 null或配合?:做空判断处理
     */
    protected var mLayoutStatusView:MultipleStatusView?=null

    /**
     * onCreate
     * override 重写标识
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(getLayoutId())
        initData()
        initView()
        start()
        initListener()
    }

    /**
     * 初始化监听
     * 函数定义关键字fun 参数格式：参数:类型
     * 此方法无返回值可以用:Unit加在后面也可以不写
     */
    private fun initListener(){
        mLayoutStatusView?.setOnClickListener(mRetryClikListener)
    }

    /**
     * 如果这个类要被修饰需要用open修饰
     */
    open val mRetryClikListener: View.OnClickListener = View.OnClickListener {
        start()
    }


    /**
     * 请求
     */
    abstract fun start()

    /**
     * 初始化View
     */
    abstract fun initView()

    /**
     * 初始化数据
     */
    abstract fun initData()

    /**
     * 获取布局ID
     * 返回值用：加类型
     */
    abstract fun getLayoutId():Int
    /**
     * 打卡软键盘
     */
    fun openKeyBord(mEditText: EditText, mContext: Context) {
        val imm = mContext.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.showSoftInput(mEditText, InputMethodManager.RESULT_SHOWN)
        imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, InputMethodManager.HIDE_IMPLICIT_ONLY)
    }
    /**
     * 关闭软键盘
     */
    fun closeKeyBord(mEditText: EditText, mContext: Context) {
        val imm = mContext.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(mEditText.windowToken, 0)
    }
}
